File: Data Structures & Algorithms/insert-into-a-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* insertIntoBST(TreeNode* root, int val) {
        if(root==nullptr)
        {
            root = new TreeNode(val);
            return root;
        }
        TreeNode* temp = root;
        recurse(temp, val);
        return temp;
    }

    void recurse(TreeNode* temp, int v)
    {
        if(v>temp->val && temp->right == nullptr)
        {
            temp->right = new TreeNode(v);
            return;
        }
        if(v<temp->val && temp->left == nullptr)
        {
            temp->left = new TreeNode(v);
            return;
        }
        else 
        {
            if(v>temp->val)
            {
                recurse(temp->right, v);
            }
            else
            {
                recurse(temp->left, v);
            }
        }
    }
};

File: Data Structures & Algorithms/lowest-common-ancestor-of-a-binary-tree-iii/submission-1.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node parent;
};
*/

class Solution {
    public Node lowestCommonAncestor(Node p, Node q) {
        if(p.parent == q|| q.parent == p)
        {
            return p.parent == q ? q : p;
        }
        if(p.parent== q.parent)
        {
            return p.parent;
        }
        Node pParent = null;
        if(p.parent!=null)
        {
            pParent = lowestCommonAncestor(p.parent, q);
        }
        Node qParent = null;
        if(q.parent!=null)
        {
            qParent = lowestCommonAncestor(p, q.parent);
        }
        return qParent == null ? pParent : qParent;
    }
}